### ISSUE TITLE
Facilitate plugin-system and extendability of the Document class

### ISSUE THREAD

**Issue by funkyfuture** [funkyfuture/delb#3]
while `delb` allows an application to extend the handling of any input to get a document out of it, it doesn't allow to extend the document class with interfaces it may need. E.g. a document loader may load from a database, but the resulting instance wouldn't have methods to store it back there.

it'd be therefore helpful to provide two [entrypoints](https://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins) that

a) can extend the document loaders
b) can return mixin-classes for the `Document` class that extend it globally

in order to bind relevant objects or set other properties to new document instances, the document loaders would also return optional keyword arguments for initializations.

@03b8, this leads me to think of `snakesist.NodeResource` as a (sub)document, that hence could be encapsuled in a `delb.Document` instance. but i'm sick and unconcentrated to elaborate more atm.

**Comment by 03b8** [comment]
encapsulating `NodeResource` in a delb `Document` sounds very reasonable. i'm also unable to look into it right now, but i will get back to you about it as soon as i can.

i wish you a speedy recovery!

**Comment by funkyfuture** [comment]
meanwhile i don't think that `pluggy`'s capabilities don't fit that well with the use here. an own implementation of a plugin system that is restricted to the actual needs is doable and should simplify the work for plugin developers and the one who's writing the documentation. this is what i would aim for:

```python
from delb.plugins.contrib.core_loaders import ftp_http_loader
from delb.plugins.contrib.https_loader import https_loader


@document_extension
class MyDocumentExtension
    ...


@register_document_loader()
def other_loader(source, config):
    ...


@register_document_loader(after=other_loader, before=https_loader)
def also_a_loader(source, config):
    ...
```

i'm not sure though whether the `after` and `before` arguments should be mutually exclusive, because i don't see the need for both now and it'd be much easier to implement.

**Comment by funkyfuture** [comment]
implemented w/ the `0.2` release, thanks @03b8 for all the support!
